Python/adaptive.md

```diff
@@ -0,0 +1,97 @@
+
+`Tadaptive = (sum(Ti*Ni) + Tmiss*Nmiss)/(sum(Ni)+Nmiss)`
+
+`Ti` is the time to execute the `i`th instruction in the family and `Ni` is the number of times that instruction is executed.
+`Tmiss` is the time to process a miss, including de-optimzation and the time to execute the base instruction.
+
+The ideal situation is where misses are rare and the specialized forms are much faster than the base instruction.
+`LOAD_GLOBAL` is near ideal, `Nmiss/sum(Ni) ≈ 0`.
+In which case we have `Tadaptive ≈ sum(Ti*Ni)`.
+Since we can expect the specialized forms `LOAD_GLOBAL_MODULE` and `LOAD_GLOBAL_BUILTIN` to be much faster than the adaptive base instruction, we would expect the specialization of `LOAD_GLOBAL` to be profitable.
+
+## Design considerations
+
+While `LOAD_GLOBAL` may be ideal, instructions like `LOAD_ATTR` and `CALL_FUNCTION` are not.
+For maximum performance we want to keep `Ti` low for all specialized instructions and `Nmiss` as low as possible.
+
+Keeping `Nmiss` low means that there should be specializations for almost
+all values seen by the base instruction. Keeping `sum(Ti*Ni)` low means keeping `Ti`
+low which means minimizing branches and dependent memory accesses (pointer chasing).
+These two objectives may be in conflict, requiring judgement and experimentation to
+design the family of instructions.
+
+### Gathering data
+
+Before choosing how to specialize an instruction, it is important to gather some data. What are the pattern of usage of the base instruction?
+Data can best be gathered by instrumenting the interpreter.
+Since a specialization function and adaptive instruction are going to be required,
+instrumentation can most easily be added in the specialization function.
+
+### Choice of specializations
+
+The performance of the specializing adaptive interpreter relies on the quality of
+specialization and keeping the overhead of specialization low.
+
+Specialized instructions must be fast. In order to be fast, specialized instructions should be tailored 
+for a particular set of values that allows them to:
+1. Verify that incoming value is part of that set with low overhead.
+2. Perform the operation quickly.
+
+This requires that the set of values is chosen such that membership can be tested quickly and
+that membership is sufficient to allow the operation to performed quickly.
+
+For example, `LOAD_GLOBAL_MODULE` is specialized for `globals()` dictionaries that have a keys with the expected version.
+
+This can be tested quickly:
+* `globals->keys->dk_version == expected_version`
+
+and the operation can be performed quickly:
+* `value = globals->keys->entries[index].value`.
+
+Because it is impossible to measure the performance of an instruction without also
+measuring unrelated factors, the assessment of the quality of a specialization will require some judgement.
+
+As a general rule, specialized instructions should be much faster than the base instruction.
```

Comment:
Just a question: how `much` faster exactly? Or is this evaluated case-by-case using our own judgement?

Reply:
It depends, so yes use your judgement.

If a specialization covers a lot of cases and reduces the number of misses by a lot, then a x2 speedup would be fine. For a more narrow specialization, something more like x4 would be appropriate.

I'm talking about the time spent in the instruction only, so for call instructions don't count the time spent in the callee or the overhead of making a C call.